Enable experimental support for the Test Explorer

marcomorain opened the pull request:
This is currently disabled behind the `calva.useTestExplorer` flag while
we test it.

Enable the Test Explorer view. This is a first pass and there are
missing features:

- [x] Re-run support
- [x] The UI that pops up is jarring - can we collapse it?
- [x] Test discovery using clojure-lsp
- [x] Assertions have the name '0'

Fixes #953
Fixes #1438
Fixes #1163

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [ ] Added to or updated docs in this branch, if appropriate
- [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. NB: *You need to sign in/up at Circle CI to find the _Artifacts_ tab.*
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [x] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe

marcomorain committed:
Enable experimental supprot for the Test Explorer

This is currently disabled behind the `calva.useTestExplorer` flag while
we test it.

Enable the Test Explorer view. This is a first pass and there are
missing features:

- [ ] Re-run support
- [ ] The UI that pops up is jarring - can we collapse it?
- [ ] Test discovery using clojure-lsp
- [ ] Assertions have the name '0'

Fixes: #953

marcomorain committed:
Enable experimental supprot for the Test Explorer

This is currently disabled behind the `calva.useTestExplorer` flag while
we test it.

Enable the Test Explorer view. This is a first pass and there are
missing features:

- [ ] The UI that pops up is jarring - can we collapse it?
- [ ] Test discovery using clojure-lsp
- [ ] Assertions have the name '0'

Fixes: #953

Enable re-test

First pass at re-run support

bpringe commented:
I did a little testing and I think this is probably fine to go out as-is with the feature flag disabled by default. Then maybe with a little promotion it can get some early adopters. @PEZ if you agree, we can merge.

PEZ commented:
Totally agree!

marcomorain committed:
Enable re-test

marcomorain committed:
First pass at re-run support

bpringe commented:
@marcomorain Are you good with us merging? I noticed you made a couple commits recently, so I want to make sure it's in a good state for merging still.

marcomorain committed:
Merge branch 'use-test-explorer' into test-tree

marcomorain committed:
Merge pull request #1 from marcomorain/test-tree

Enable experimental supprot for the Test Explorer

marcomorain committed:
Add test types

marcomorain committed:
Merge branch 'dev' into use-test-explorer

marcomorain committed:
Add some polish

marcomorain commented:
@bpringe I'm going to get this into a more polished state today hopefully.

marcomorain committed:
Merge remote-tracking branch 'origin/dev' into use-test-explorer

marcomorain committed:
Merge remote-tracking branch 'origin/dev' into use-test-explorer

marcomorain commented:
@bpringe @PEZ I've gone through the code, and I'm happy with it. I've also gone through the checklist and I think this is all good to merge to dev 🚀 

The functionality is all behind a feature flag at present, so the change should have zero impact unless a user opts-in. If a user has not enabled the Test Explorer, we show test errors in the output Window, and using Diagnostics. If the user does enable the new system, then test errors will continue to be shown in the output Window, but the use of Diagnostics is replaced with the Test Controller API (populating the Test Explorer UI).

PEZ commented:
You are great, @marcomorain! I'll run with this enabled a bit.

As @bpringe mentioned on Slack, this would be good to have documented in the docs. Here: https://calva.io/test-runner/ So that people can easier know if they should enable the experiment or not, and what the prerequisites are, and what to expect when it's enabled.

PEZ commented:
We can merge w/o the docs, though. And add them quickly afterwards. I like to write those so if you rather want me to do it, @marcomorain , I'm game.

Before merging we will need to fix the conflict. And I will have a look at the code, hopefully @bpringe has some time for that as well.

PEZ commented on src/extension.ts:
```diff
@@ -86,7 +86,14 @@ async function activate(context: vscode.ExtensionContext) {
-    lsp.activate(context);
+
+
+    const controller = vscode.tests.createTestController('calvaTestController', 'Clojure Cider')
```
Does `Clojure Cider` show up in the UI somewhere? Or what is it for?

marcomorain replied:
Good question. The first two arts are a globally unique ID and a human readable label. I haven’t seen the label in the UI, but I’ve just noticed that the name that I doesn’t have the word “Calva” in it.

marcomorain replied:
Here is the Go extension:

https://github.com/golang/vscode-go/blob/944fbf7185aa47cd3832bea776420996ae599e89/src/goTest/explore.ts#L37

PEZ replied:
Yes, so that's why I am asking. It's probably best to give the label `Calva`. Wherever it shows up, that will make it clear  where it comes from. =) Not a blocker for merge though, imo.

PEZ commented on src/lsp/types.ts:
```diff
@@ -0,0 +1,37 @@
+    tree: TestTreeNode
+}
+
+// The Clojure LSP client will call this handler function any time a test is
+// detected. This allows Calva to populate the Test Explorer from the data provided
+// by Clojure LSP. This indirection allows us to write testRunner.ts without
+// specific knowledge of LSP, and also write the LSP client without any specific
+// knowledge of the testRunner code.
```
Brilliant!

PEZ commented on src/testRunner.ts:
```diff
@@ -5,13 +5,154 @@ import { disabledPrettyPrinter } from './printer';
+// - replacing . with /
+// - replace - with _
+// - search for .clj or cljs files that match.
+async function uriForNamespace(namespace: string): Promise<vscode.Uri> {
+    const path = namespace.replace(/\./g, "/").replace(/-/g, "_");
+    const uris = await vscode.workspace.findFiles('**/' + path + ".{clj,cljs}", null, 1);
+    return uris[0];
+}
```
This might be a bit brittle, especially in a monorepo. If we assume we have a REPL connection, it's probably better to rely on nrepl for this. Maybe factor the `getUriForCurrentNamespace` out of `results-doc.ts` moving it to `namespace.ts`:

https://github.com/BetterThanTomorrow/calva/blob/60bb98616b577a960f15473a821978b62782ca9b/src/results-output/results-doc.ts#L374-L375

marcomorain replied:
I've done this 👍

PEZ reviewed (commented):
This is so awesome. Thanks!

I'd like the namespace->uri thing to be considered and also wonder about where `Clojure Cider` might show to the user.

marcomorain commented:
Thanks

> This is so awesome. Thanks!
> 
> I'd like the namespace->uri thing to be considered and also wonder about where `Clojure Cider` might show to the user.

Thanks, I’ll take care of these today.

PEZ commented:
How much are we relying on clojure-lsp here, btw? We are considering giving the user's the option to disable clojure-lsp and then we need to gracefully degrade things that rely on it.

marcomorain commented:
There is no reliance on LSP - things will work without it.

using LSP gives two benefits:
- it adds line numbers to the test tree nodes for passed tests. Cider only reports line numbers for failed tests. So with LSP you can click on a passed test to view the test. Without LSP the passed tests are clickable. 
- LSP discovers tests before you run them. Without LSP, Test items will only appear in the Explorer after you run tests in a file containing tests. With LSP the tests appear when you open a file containing tests.

PEZ commented:
Super awesome!

marcomorain committed:
Merge branch 'use-test-explorer' of github.com:marcomorain/calva into use-test-explorer

marcomorain committed:
Extract getUriForNamespace into namespace.ts

Extract getUriForNamespace and call it from the testRunner namespace
when required.

marcomorain committed:
Merge branch 'dev' into use-test-explorer

marcomorain committed:
Add some docs

marcomorain committed:
Merge branch 'use-test-explorer' of github.com:marcomorain/calva into use-test-explorer

marcomorain commented:
@PEZ I've made those changes and added some docs @ https://github.com/BetterThanTomorrow/calva/pull/1398/commits/2650e970b208e182a279f43f1712cc2606834d19#diff-0f3ec18fe95a0fdcd28d3c88e1af3a664c866d6c57e6ef8612f4ccb8b84b39aaR27

PEZ commented on docs/site/test-runner.md:
```diff
@@ -24,6 +24,14 @@ Toggle between implementation and test | - | Switches the file between implement
 
+## Test Explorer
+
+Calva has experimental support for showing test results in the VSCode Test Explorer UI. You can enable this support by setting `calva.useTestExplorer` to `true`. When you enable this setting, the Testing icon will appear in the VSCode Activilty Bar.
+
+With this feature you will be able to browse and run tests directly from the Testing tab.
```
I think we should add something about what the Test Explorer is. Or maybe just link Test Explorer UI to its marketplace page.

marcomorain replied:
Shall we get this merged and pull the docs changes out into a separate PR?

PEZ replied:
```suggestion
Calva has experimental support for showing test results in the [VSCode Test Explorer UI](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer). You can enable this support by setting `calva.useTestExplorer` to `true`. When you enable this setting, the Testing icon will appear in the VSCode Activilty Bar.

With this feature (and the TestExplorer UI extension installed) you will be able to browse and run tests directly from the Testing tab.
```

PEZ committed:
Update docs/site/test-runner.md

Add testexplorer mention

PEZ reviewed (approved):
LGTM. Let's merge!

PEZ commented:
🎉

athos commented:
Hi, thank you for working on this amazing feature!

By the way, does this feature really depend on the [Test Explorer UI](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer) extension (as mentioned in `docs/site/test-runner.md`)? As far as I can see from the code change, it seems to depend only on VS Code's native [Testing API](https://code.visualstudio.com/api/extension-guides/testing). Am I missing something?

PEZ commented:
You might be right. I assumed test explorer ui extension, which I have installed since i use it heavily for Calva dev.
